### PR TITLE
feat: used go-pretty to improve table formatting on screen width

### PR DESCRIPTION
**Pull request by Aerex**
# Context

The purpose of this PR is to improve the table formatting on the columns to wrap when the text when it reaches an estimated max column width. The max col width is calculated by excluding the padding and applying the remaining space evenly among the columns. Any columns that are expected to be wider than the others (eg. **ID** and **Description**) will given more space than the others

The PR uses the [go-pretty](https://github.com/jedib0t/go-pretty) library to manage the column text wrapping, and max column width and coloring.

### CODE REVIEW

File: bluemix/terminal/table_test.go [hunk 1]
```diff
@@ -38,7 +38,7 @@ func TestEmptyHeaderTable(t *testing.T) {
 	testTable.Add("row1", "row2")
 	testTable.Print()
 	assert.Contains(t, buf.String(), "row1")
-	assert.Equal(t, "       \nrow1   row2\n", buf.String())
+	assert.Equal(t, "\nrow1   row2\n", buf.String())
```
Comment (Aerex) [inline bluemix/terminal/table_test.go:41]: **Callout**: Removing the tab space does not change the output visually. 

image

File: bluemix/terminal/table_test.go [hunk 2]
```diff
@@ -79,7 +79,7 @@ func TestNotEnoughRowEntires(t *testing.T) {
 	testTable.Add("", "row2")
 	testTable.Print()
 	assert.Contains(t, buf.String(), "row1")
-	assert.Equal(t, "col1   col2\nrow1   \n       row2\n", buf.String())
+	assert.Equal(t, "col1   col2\nrow1\n       row2\n", buf.String())
```
Comment (Aerex) [inline bluemix/terminal/table_test.go:82]: **Callout**: Same as above

image

**Comment by Aerex** [comment]
Ready for review @steveclay

File: bluemix/terminal/table.go [hunk 3]
```diff
@@ -69,58 +75,152 @@ func (t *PrintableTable) Add(row ...string) {
...
-}
 
-func (t *PrintableTable) printRow(row []string) {
-	output := ""
-	for columnIndex, value := range row {
-		if columnIndex == 0 {
-			value = TableContentHeaderColor(value)
+	for i := range t.rows {
+		var row, emptyRow table.Row
+		for i, cell := range t.rows[i] {
```
Comment (steveclay) [inline bluemix/terminal/table.go:201]: Variable name `i` is used twice in this context. Let's also rename these so that they indicate what they are doing. For example, the 1st i is a rowIndex and the 2nd a cellIndex it appears.

File: bluemix/terminal/table.go [hunk 4]
```diff
@@ -69,58 +75,152 @@ func (t *PrintableTable) Add(row ...string) {
...
+
+	colCount := len(t.rows[0])
+	var (
+		widestColIndicies []int
+		terminalWidth     = terminalWidth()
+		// total amount padding space that a row will take up
+		totalPaddingSpace = (colCount - 1) * minSpace
+		remainingSpace    = terminalWidth - totalPaddingSpace
+		// the estimated max column width by dividing the remaining space evenly across the columns
+		maxColWidth = (terminalWidth - totalPaddingSpace) / colCount
```
Comment (steveclay) [inline bluemix/terminal/table.go:145]: This is repeating the calculation for remainingSpace.

File: bluemix/terminal/table.go [hunk 5]
```diff
@@ -69,58 +75,152 @@ func (t *PrintableTable) Add(row ...string) {
...
+		// using the remaining space
+		if t.maxSizes[i] > maxColWidth && (len(t.headers) > 0 && isWideColumn(t.headers[i])) {
+			widestColIndicies = append(widestColIndicies, i)
+		} else if t.maxSizes[i] < maxColWidth {
+			// use the max column width instead of the estimated max column width
+			// if it is shorter
+			columnConfig[i].WidthMax = t.maxSizes[i]
+			remainingSpace -= t.maxSizes[i]
+		} else {
+			remainingSpace -= maxColWidth
```
Comment (steveclay) [inline bluemix/terminal/table.go:167]: Wondering if there can be a scenario that remainingSpace becomes 0 or less.

File: bluemix/terminal/table.go [hunk 6]
```diff
@@ -69,58 +75,152 @@ func (t *PrintableTable) Add(row ...string) {
...
+	headerRow, rows := t.createPrettyRowsAndHeaders()
+	columnConfig := t.createColumnConfigs()
+
+	tbl.SetColumnConfigs(columnConfig)
+	tbl.AppendHeader(headerRow)
+	tbl.AppendRows(rows)
+	tbl.Render()
+}
+
+func (t *PrintableTable) createColumnConfigs() []table.ColumnConfig {
```
Comment (steveclay) [inline bluemix/terminal/table.go:131]: It would be good to add some more testing to explicitly try out some odd column headings to ensure there's no breaking conditions. I don't have any in mind, but would like to see if we can test some boundary conditions here.

File: bluemix/terminal/table.go (hunk 3 above)
Comment (Aerex) [reply bluemix/terminal/table.go:201]: Yeah good catch. Let me update

File: bluemix/terminal/table.go [hunk 7]
```diff
@@ -69,58 +75,152 @@ func (t *PrintableTable) Add(row ...string) {
...
+		columnConfig[i] = table.ColumnConfig{
+			AlignHeader: text.AlignLeft,
+			Align:       text.AlignLeft,
+			WidthMax:    maxColWidth,
+			Number:      i + 1,
+		}
+
+		// assuming the table has headers: store columns with wide content where the max width may need to be adjusted
+		// using the remaining space
+		if t.maxSizes[i] > maxColWidth && (len(t.headers) > 0 && isWideColumn(t.headers[i])) {
```
[the remark this answers is not in the document]
Comment (Aerex) [reply bluemix/terminal/table.go:159]: I don't follow. Are you saying that adding `len(t.headers) > 0` will not prevent calling `t.headers[i]` when calling `isWideColumn`? 

If so, I have tested that scenario in the unit tests where the headers is not defined.

File: bluemix/terminal/table.go (hunk 5 above)
Comment (Aerex) [reply bluemix/terminal/table.go:167]: Good question. I don't think there is a valid scenario since that would mean the user would need to have a lot of columns. For instance, for terminal width of 80, the user would need to create more than 27 columns (27 * 3 > 80) for there to be negative `remainingSpace`.

File: bluemix/terminal/table.go (hunk 6 above)
Comment (Aerex) [reply bluemix/terminal/table.go:140]: Added more unit tests for various scenarios.

File: bluemix/terminal/table.go [hunk 8]
```diff
@@ -69,58 +76,160 @@ func (t *PrintableTable) Add(row ...string) {
...
+func terminalWidth() int {
+	var err error
+	terminalWidth, _, err := term.GetSize(int(os.Stdin.Fd()))
+
+	if err != nil {
+		// Assume normal 80 char width line
+		terminalWidth = 80
+	}
+
+	testTerminalWidth, envSet := os.LookupEnv("TEST_TERMINAL_WIDTH")
```
Comment (Aerex) [inline bluemix/terminal/table.go:102]: **Callout**: Added test environment variable to test variable terminal width in unit tests

File: bluemix/terminal/table.go [hunk 9]
```diff
@@ -69,58 +76,160 @@ func (t *PrintableTable) Add(row ...string) {
...
 
-	for _, line := range t.rows {
-		t.printRow(line)
+	colCount := len(t.rows[0])
+	var (
+		widestColIndicies []int
+		terminalWidth     = terminalWidth()
+		// total amount padding space that a row will take up
+		totalPaddingSpace = (colCount - 1) * minSpace
+		remainingSpace    = max(0, terminalWidth-totalPaddingSpace)
```
[the remark this answers is not in the document]
Comment (steveclay) [reply bluemix/terminal/table.go:152]: Hm, with this the next line could end up with a maxColWidth of 0 (0 / colCount)

File: bluemix/terminal/table.go [hunk 10]
```diff
@@ -156,7 +165,7 @@ func (t *PrintableTable) createColumnConfigs() []table.ColumnConfig {
 
 		// assuming the table has headers: store columns with wide content where the max width may need to be adjusted
 		// using the remaining space
-		if t.maxSizes[i] > maxColWidth && (len(t.headers) > 0 && isWideColumn(t.headers[i])) {
+		if t.maxSizes[i] > maxColWidth && (i < len(t.headers) && isWideColumn(t.headers[i])) {
```
Comment (Aerex) [inline bluemix/terminal/table.go:168]: Change `i` to `colIndex` to better understand what `i` is

File: bluemix/terminal/table.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (steveclay) [reply bluemix/terminal/table.go:152]: Discussed during code review. This should be fine from our walkthrough

**Review by steveclay — APPROVED** [review]
